Data/HashMap/Base.hs

```diff
@@ -209,6 +212,19 @@ instance Foldable.Foldable (HashMap k) where
     {-# INLINE length #-}
 #endif
 
+#if MIN_VERSION_base(4,10,0)
+-- | @since UNRELEASED
+instance Bifoldable HashMap where
+    bifoldr f g = go
+      where
+        go z Empty                 = z
+        go z (Leaf _ (L k v))      = f k (g v z)
+        go z (BitmapIndexed _ ary) = A.foldr (flip go) z ary
+        go z (Full ary)            = A.foldr (flip go) z ary
+        go z (Collision _ ary)     = A.foldr (\ (L k v) z' -> f k (g v z')) z ary
+    {-# INLINE bifoldr #-}
+#endif
+
```

Comment:
Why not just use `foldrWithKey` to define `bifoldr`? We should also define `bifoldMap` following the tree structure.

Reply:
Also `foldlWithKey`, etc.

Reply:
While we're at it, we definitely want `foldMapWithKey`, like we have for `Data.Map`. This can be used to define the rather less useful `bifoldMap`.

Reply:
@treeowl You're already adding `foldMapWithKey` in #224. Do we expect that PR to merge anytime soon, or should I duplicate pieces of it into this one?

Reply:
Well that was a fast merge. Anyway, rebased and done.